### PR TITLE
Feature/improve generic oauth

### LINKED ISSUE #6176: incomplete implementation of generic_oauth support
(follow up on #4718)
* **I'm submitting a ...**
- [ ] Bug report
- [X] Feature request
- [ ] Question / Support request: **Please do not** open a github issue. [Support Options](http://grafana.org/support/)

Please include this information:
- What Grafana version are you using?
4.0.0 pre1 (grabbed #5899 and rebased off of origin/master HEAD)
- What datasource are you using?
influxdb and elasticsearch, but that's not exactly relevant
- What OS are you running grafana on?
CentOS 7
- What did you do?
Tested generic_oauth support (#4718) on a custom service that supports the oauth2 standard and requires TLS auth and `state` query param set
- What was the expected result?
That grafana would properly authenticate against my service
- What happened instead?
My oauth service requires `state` query param to be set (there currently isn't a way to configure this in the existing implementation), so that service was rejecting grafana's attempts to redirect. I was able to workaround this by taking the URL grafana returned to the browser and adding a dummy value for `state` to the query param, which then led me to a bug/FR, where grafana threw a 500 error b/c of TLS trust issues: ```t=2016-10-03T16:41:48-0400 lvl=eror msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/login/generic_oauth status=500 remote_addr=10.217.68.240 time_ms=13ns size=709```

I'm going to take a stab at fixing this myself and opening a PR

### PR DESCRIPTION
**Pull request by tomkozlowski**
Fixes #6176

### CODE REVIEW

File: pkg/api/login_oauth.go [hunk 1]
```diff
@@ -29,20 +35,49 @@ func OAuthLogin(ctx *middleware.Context) {
...
+				InsecureSkipVerify: true,
+				Certificates: []tls.Certificate{cert},
+				RootCAs: caCertPool,
+				},
+	}
+	sslcli := &http.Client{Transport: tr}
+	oauthCtx := context.TODO()
+	oauthCtx = context.WithValue(oauthCtx, oauth2.HTTPClient, sslcli)
+
+	//TODO: add check to see if tls_client_* aren't all set, set oauthCtx to oauth2.NoContext
```
Comment (DanCech) [inline pkg/api/login_oauth.go]: This will definitely be needed before this can be merged

File: pkg/api/login_oauth.go [hunk 2]
```diff
@@ -56,14 +91,7 @@ func OAuthLogin(ctx *middleware.Context) {
...
-	// validate that the email is allowed to login to grafana
-	if !connect.IsEmailAllowed(userInfo.Email) {
-		ctx.Logger.Info("OAuth login attempt with unallowed email", "email", userInfo.Email)
-		ctx.Redirect(setting.AppSubUrl + "/login?failCode=1002")
-		return
-	}
-
-	userQuery := m.GetUserByLoginQuery{LoginOrEmail: userInfo.Email}
+	userQuery := m.GetUserByLoginQuery{LoginOrEmail: userInfo.Identity}
 	err = bus.Dispatch(&userQuery)
```
Comment (DanCech) [inline pkg/api/login_oauth.go]: This will also break existing implementations

**Comment by DanCech** [comment]
It looks like you've made some rather sweeping changes to this module, which would break anyone currently using it.

**Comment by tomkozlowski** [comment]
Yeah, I was hoping to get some suggestions on how to maintain backcompat. The usage of `email` as an indexed, unique key (rather than something like Identity or Name) is kinda frustrating and limiting

**Comment by DanCech** [comment]
I've been doing some work on this stuff, let me take a deeper look tomorrow and see what I can come up with.  I've been thinking about how to make the separate endpoints for getting user email(s), team(s) and org(s) anyway.

Can you provide an example of the userInfo your provider is returning?

**Comment by tomkozlowski** [comment]
sure @DanCech. here's the JSON output from our `api_url` output:
```
{
  "id": "8d09f8a0-5eb9-4e32-be7a-5fb2bd567ea1",
  "username": "tom",
  "attributes": {
    "email:primary": [
      "tom@gmail.com"
    ],
    "realm": [
      "internal-realm"
    ]
  }
}
```

**Comment by torkelo** [comment]
thanks for contributing & wanting to help make Grafana better. looks like a good start for making the generic oauth more generic (and follow spec). Will needs some changes so it doesn't break other oauth implementations.

**Comment by DanCech** [comment]
@tomkozlowski please give the code in #6226 a try, it should work with your API while maintaining backward compatibility.

**Comment by torkelo** [comment]
I think https://github.com/grafana/grafana/pull/6226  addresses most of this PR (and some of the changes in https://github.com/grafana/grafana/pull/6226 was inspired I think by your great work started in this PR)

**Comment by torkelo** [comment]
ah, missed your comment @DanCech

**Comment by tomkozlowski** [comment]
@DanCech @torkelo yeah, I think Dan's PR #6226 looks a bit better for maintaining backcompat, however, I think it still enforces some GitHub concepts as generic OAuth concepts (I left comments on the merged PR)

**Pull request closed without merging**